rhel/vex/fetcher.go

```diff
@@ -0,0 +1,388 @@
+		if err != nil {
+			return nil, hint, err
+		}
+		latestReq, err := http.NewRequestWithContext(ctx, http.MethodGet, latestURI.String(), nil)
+		if err != nil {
+			return nil, hint, err
+		}
+		latestRes, err := u.client.Do(latestReq)
+		if err != nil {
+			return nil, hint, err
+		}
+		defer latestRes.Body.Close()
+
+		if latestRes.StatusCode != http.StatusOK {
+			return nil, hint, fmt.Errorf("unexpected response from archive_latest.txt: %s", latestRes.Status)
+		}
+
+		body, err := io.ReadAll(latestRes.Body) // Fine to use as expecting small number of bytes.
+		if err != nil {
+			return nil, hint, err
+		}
+
+		compressedFilename := string(body)
+		zlog.Debug(ctx).
+			Str("filename", compressedFilename).
+			Msg("requesting latest compressed file")
+
+		uri, err := u.url.Parse(compressedFilename)
+		if err != nil {
+			return nil, hint, err
+		}
+
+		rctx, cancel := context.WithTimeout(ctx, compressedFileTimeout)
+		defer cancel()
+
+		req, err := http.NewRequestWithContext(rctx, http.MethodGet, uri.String(), nil)
+		if err != nil {
+			return nil, hint, err
+		}
+
+		res, err := u.client.Do(req)
+		if res != nil {
+			defer res.Body.Close()
+		}
+		if err != nil {
+			return nil, hint, err
+		}
+		if res.StatusCode != http.StatusOK {
+			return nil, hint, fmt.Errorf("unexpected response from latest compressed file: %s", res.Status)
+		}
+
+		lm := res.Header.Get("last-modified")
+		fp.requestTime, err = time.Parse(http.TimeFormat, lm)
+		if err != nil {
+			return nil, hint, fmt.Errorf("could not parse last-modified header %s: %w", lm, err)
+		}
+		z, err := zstd.NewReader(res.Body)
+		if err != nil {
+			return nil, hint, err
+		}
+		defer z.Close()
+		r := tar.NewReader(z)
+
+		var (
+			h              *tar.Header
+			buf, bc        bytes.Buffer
+			entriesWritten int
+		)
+		for h, err = r.Next(); errors.Is(err, nil); h, err = r.Next() {
+			if h.Typeflag != tar.TypeReg {
+				continue
+			}
+			year, err := strconv.ParseInt(path.Dir(h.Name), 10, 64)
+			if err != nil {
+				return nil, hint, fmt.Errorf("error parsing year %w", err)
+			}
+			if year < lookBackToYear {
+				continue
+			}
+			buf.Grow(int(h.Size))
+			if _, err := buf.ReadFrom(r); err != nil {
+				return nil, hint, err
+			}
+
+			err = json.Compact(&bc, buf.Bytes())
+			if err != nil {
+				return nil, hint, fmt.Errorf("error compressing JSON %s: %w", h.Name, err)
+			}
+			bc.WriteByte('\n')
+			cw.Write(bc.Bytes())
+			buf.Reset()
+			bc.Reset()
+			entriesWritten++
+		}
+		if !errors.Is(err, io.EOF) {
+			return nil, hint, fmt.Errorf("error reading tar contents: %w", err)
+		}
+
+		zlog.Debug(ctx).
+			Str("updater", u.Name()).
+			Int("entries written", entriesWritten).
+			Msg("finished writing compressed data to spool")
+	}
+
+	err = u.processChanges(ctx, cw, fp)
+	if err != nil {
+		return nil, hint, err
+	}
+
+	err = u.processDeletions(ctx, cw, fp)
+	if err != nil {
+		return nil, hint, err
+	}
+
+	fp.requestTime = time.Now()
```

Comment:
If the compressed file is processed (i.e. the `changesEtag` is empty) we want to process all the changes that happened since the compressed file was modified (i.e. `fp.requestTime, err = time.Parse(http.TimeFormat, lm`). If the compressed file was not read (i.e. this wasn't the first time the updater was run) We want to process all the changes since the last time we processed the changes i.e. setting `fp.requestTime = time.Now()` so the next run knows where it left off. Let me know if this makes sense.